torch/_dynamo/backends/train_step.py

```diff
@@ -0,0 +1,167 @@
+from contextlib import contextmanager
+from copy import copy
+from typing import Any, Dict, List
+
+import torch
+import torch.utils._pytree as pytree
+from torch import fx
+from torch._dynamo import register_backend
+from torch._dynamo.backends.registry import lookup_backend
+from torch._subclasses.fake_tensor import FakeTensorMode
+
+from torch.func import functionalize
+from torch.fx.experimental.proxy_tensor import make_fx
+from torch.fx.interpreter import Interpreter
+from torch.nn.utils import stateless
+
+
+@contextmanager
+def _rematerialize_optimizer(
+    opt: torch.optim.Optimizer,
+    named_states: Dict[str, Any],
+    params: Dict[str, torch.nn.Parameter],
+):
+    assert opt is not None
+
+    # update opt.state with proxy tensors
+    orig_states: Dict[str, Any] = copy(opt.state)
+    for n in named_states:
+        # opt.state's key type is string, but optimizer uses Parameter as keys
+        opt.state[params[n]] = named_states[n]  # type: ignore[index]
+
+    # FIXME: support multiple parameter groups
+    param_group = opt.param_groups[0]
+    orig_params = param_group["params"]
+    # FIXME(@mrshenli): exclude buffers
+    param_group["params"] = params.values()
+
+    try:
+        yield
+    finally:
+        param_group["params"] = orig_params
+        opt.state.update(orig_states)
+
+
+def train_step_compiler(backend_compile_fn):
+    """Note [Train Step Compile]
+
+    Usually, torch.compile() allows graph-breaks and compiles pairs of forward (+backward) by
+    extracting sections of forward from python programs and using AotAutograd to produce corresponding
+    chunks of backwards, tying it back together with an AotFunction.
+
+    Instead, TrainStepCompiler assumes the user compiles a full train_step function complete with calls to
+    .backward(), optimizer step(), and zero_grad().  It additionally requires no graph-breaks.
```

Comment:
Note that we want to check that the user doesn't pass `set_to_none=False` to this function to ensure there is no side effect that we wouldn't replicate in the original function.